lib/posix/timer.c

```diff
@@ -77,52 +110,89 @@ static void zephyr_timer_interrupt(struct k_timer *ztimer)
  */
 int timer_create(clockid_t clockid, struct sigevent *evp, timer_t *timerid)
 {
+	int ret = 0;
+	int detachstate;
 	struct timer_obj *timer;
 	const k_timeout_t alloc_timeout = K_MSEC(CONFIG_TIMER_CREATE_WAIT);
 
-	if (evp == NULL || (evp->sigev_notify != SIGEV_NONE && evp->sigev_notify != SIGEV_SIGNAL &&
-			    evp->sigev_notify != SIGEV_THREAD)) {
+	if (evp == NULL || timerid == NULL) {
 		errno = EINVAL;
 		return -1;
 	}
 
-	if (k_mem_slab_alloc(&posix_timer_slab, (void **)&timer, alloc_timeout) == 0) {
-		(void)memset(timer, 0, sizeof(struct timer_obj));
-	} else {
+	if (k_mem_slab_alloc(&posix_timer_slab, (void **)&timer, alloc_timeout) != 0) {
+		LOG_DBG("k_mem_slab_alloc() failed: %d", ret);
 		errno = ENOMEM;
 		return -1;
 	}
 
-	timer->sev.sigev_notify_function = evp->sigev_notify_function;
-	timer->sev.sigev_value = evp->sigev_value;
-	timer->interval.tv_sec = 0;
-	timer->interval.tv_nsec = 0;
-	timer->reload = 0U;
-	timer->status = NOT_ACTIVE;
+	*timer = (struct timer_obj){0};
+	timer->evp = *evp;
+	evp = &timer->evp;
 
-	if (evp->sigev_notify == SIGEV_NONE) {
+	switch (evp->sigev_notify) {
+	case SIGEV_NONE:
 		k_timer_init(&timer->ztimer, NULL, NULL);
-	} else if (evp->sigev_notify == SIGEV_THREAD) {
-		int ret;
+		break;
+	case SIGEV_SIGNAL:
+		k_timer_init(&timer->ztimer, zephyr_timer_wrapper, NULL);
+		break;
+	case SIGEV_THREAD:
+		if (evp->sigev_notify_attributes != NULL) {
+			ret = pthread_attr_getdetachstate(evp->sigev_notify_attributes,
+							  &detachstate);
+			if (ret != 0) {
+				LOG_DBG("pthread_attr_getdetachstate() failed: %d", ret);
+				errno = ret;
+				ret = -1;
+				goto free_timer;
+			}
+
+			if (detachstate != PTHREAD_CREATE_DETACHED) {
+				ret = pthread_attr_setdetachstate(evp->sigev_notify_attributes,
+								  PTHREAD_CREATE_DETACHED);
+				if (ret != 0) {
+					LOG_DBG("pthread_attr_setdetachstate() failed: %d", ret);
+					errno = ret;
+					ret = -1;
+					goto free_timer;
+				}
+			}
+		}
+
+		ret = k_sem_init(&timer->sem_cond, 0, 1);
+		if (ret != 0) {
+			LOG_DBG("k_sem_init() failed: %d", ret);
+			errno = -ret;
+			ret = -1;
+			goto free_timer;
+		}
 
-		timer->sev.sigev_notify = SIGEV_THREAD;
-		k_sem_init(&timer->sem_cond, 0, 1);
 		ret = pthread_create(&timer->thread, evp->sigev_notify_attributes,
 							zephyr_thread_wrapper, timer);
 		if (ret != 0) {
-			k_mem_slab_free(&posix_timer_slab, (void *) &timer);
-			return ret;
+			LOG_DBG("pthread_create() failed: %d", ret);
+			errno = ret;
+			ret = -1;
+			goto free_timer;
 		}
 
-		pthread_detach(timer->thread);
 		k_timer_init(&timer->ztimer, zephyr_timer_interrupt, NULL);
-	} else {
-		k_timer_init(&timer->ztimer, zephyr_timer_wrapper, NULL);
+		break;
+	default:
+		ret = -1;
+		errno = EINVAL;
+		goto free_timer;
 	}
 
 	*timerid = (timer_t)timer;
+	goto out;
 
-	return 0;
+free_timer:
+	k_mem_slab_free(&posix_timer_slab, (void *)&timer);
```

Comment:
Could we place the ret and errno setting here, so something like
```
free_timer:
    errno = ret;
    ret = -1;
    k_mem_slab_free();
```
This way you do not need to set ret and errno multiple times above.

Reply:
The issue is that assigning it will differ based on whether we are failing from a kernel API or a POSIX API (pthread APIs return `errno` unsigned, while `k_` APIs return `-errno`)

Reply:
Ok, makes sense, never mind then.